Homework2/solution.cpp

```diff
@@ -0,0 +1,80 @@
+#include <iostream>
+#include <numeric>
+#include <thread>
+#include <vector>
+
+template<typename It, typename T>
+struct accumulator {
+    void operator()(It first, It last, T &result) {
+        result = std::accumulate(first, last, result);
+    }
+};
+
+void set_th_parameters(long length, unsigned int &th_quantity, long &block_size) {
+    auto hw_threads = std::thread::hardware_concurrency();
+    std::cout << "Threads used " << hw_threads << std::endl;
+    th_quantity = hw_threads != 0 ? hw_threads : 2;
+    block_size = length / th_quantity;
+}
+
+template<typename It, typename T>
+T th_accumulate(It first, It last, T init) {
+    auto length = std::distance(first, last);
+    if (!length) return init;
+
+    unsigned int th_quantity;
+    long block_size;
```

Comment:
najlepiej jest użyć domyślnej inizjalizacji za pomocą klamerek {}.
Gdyby coś rzuciło wyjątkiem to nie będzie aż tak losowych skutków ;)